File: core/src/ru/deadsoftware/cavedroid/CaveGame.java
package ru.deadsoftware.cavedroid;

import com.badlogic.gdx.Application;
import com.badlogic.gdx.Game;
import com.badlogic.gdx.Gdx;
import org.jetbrains.annotations.Nullable;
import ru.deadsoftware.cavedroid.game.GameScreen;
import ru.deadsoftware.cavedroid.misc.Assets;
import ru.deadsoftware.cavedroid.misc.utils.AssetLoader;
import ru.deadsoftware.cavedroid.prefs.PreferencesStore;

public class CaveGame extends Game {

    private static final String TAG = "CaveGame";

    public static final String VERSION = "alpha 0.9.2";

    private final MainConfig mMainConfig;
    private final MainComponent mMainComponent;
    private final AssetLoader mAssetLoader;

    private final String mGameFolder;
    private final boolean mTouch;
    private boolean mDebug;

    @Nullable
    private final String mAssetsPackPath;

    public CaveGame(String gameFolder,
                    boolean touch,
                    PreferencesStore preferencesStore,
                    @Nullable String assetsPackPath) {
        mGameFolder = gameFolder;
        mTouch = touch;
        mAssetsPackPath = assetsPackPath;

        mMainComponent = DaggerMainComponent
                .builder()
                .caveGame(this)
                .preferencesStore(preferencesStore)
                .build();

        mMainConfig = mMainComponent.getMainConfig();
        mAssetLoader = mMainComponent.getAssetLoader();
    }

    public void setDebug(boolean debug) {
        mDebug = debug;
    }

    private void initConfig() {
        int width = 480;
        int height = (int) (width * ((float) Gdx.graphics.getHeight() / Gdx.graphics.getWidth()));

        mMainConfig.setMainComponent(mMainComponent);
        mMainConfig.setGameFolder(mGameFolder);
        mMainConfig.setTouch(mTouch);
        mMainConfig.setWidth(width);
        mMainConfig.setHeight(height);
        mMainConfig.setShowInfo(mDebug);
        mMainConfig.setAssetsPackPath(mAssetsPackPath);

        if (mDebug) {
            Gdx.app.setLogLevel(Application.LOG_DEBUG);
        } else {
            Gdx.app.setLogLevel(Application.LOG_ERROR);
        }

        mMainConfig.setFullscreenToggleListener((value) -> {
            if (value) {
                Gdx.graphics.setFullscreenMode(Gdx.graphics.getDisplayMode());
            } else {
                Gdx.graphics.setWindowedMode(width, height);
            }
        });
    }

    public void newGame(int gameMode) {
        GameScreen gameScreen = mMainComponent.getGameScreen();
        gameScreen.newGame(gameMode);
        setScreen(gameScreen);
    }

    public void loadGame() {
        GameScreen gameScreen = mMainComponent.getGameScreen();
        gameScreen.loadGame();
        setScreen(gameScreen);
    }

    public void quitGame() {
        if (screen != null) {
            screen.dispose();
        }
        setScreen(mMainComponent.getMenuScreen());
    }

    @Override
    public void create() {
        Gdx.files.absolute(mGameFolder).mkdirs();
        initConfig();

        Gdx.app.debug(TAG, mGameFolder);
        Assets.load(mAssetLoader);
        setScreen(mMainComponent.getMenuScreen());
    }

    @Override
    public void dispose() {
        if (screen != null) {
            screen.dispose();
        }
        Assets.dispose();
    }
}

File: core/src/main/kotlin/ru/deadsoftware/cavedroid/CaveGame.kt
package ru.deadsoftware.cavedroid

import com.badlogic.gdx.Application
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.math.Rectangle
import ru.fredboy.cavedroid.common.api.PreferencesStore
import ru.fredboy.cavedroid.common.utils.ratio
import ru.fredboy.cavedroid.domain.configuration.model.CameraContext
import ru.fredboy.cavedroid.domain.configuration.repository.GameContextRepository

class CaveGame(
    private val gameDataDirectoryPath: String,
    private val isTouchScreen: Boolean,
    private val isDebug: Boolean,
    private val preferencesStore: PreferencesStore,
) : BaseGame() {

    private val mainComponent: MainComponent
    private val mainConfig: MainConfig

    private val mGameContextRepository: GameContextRepository

    init {
        mainComponent = DaggerMainComponent.builder()
            .caveGame(this)
            .preferencesStore(preferencesStore)
            .build()

        mainConfig = mainComponent.mainConfig
        mGameContextRepository = mainComponent.gameContextRepository
    }

    private fun initMainConfig() {
        val width = DEFAULT_VIEWPORT_WIDTH
        val height = width / Gdx.graphics.ratio

        mainConfig.mainComponent = mainComponent

        mGameContextRepository.apply {
            setGameDirectory(gameDataDirectoryPath)
            setTouch(isTouchScreen)
            setWidth(width)
            setHeight(height)
            setShowInfo(isDebug)

            setCameraContext(
                CameraContext(
                    viewport = Rectangle(0f, 0f, width, height),
                    camera = OrthographicCamera().apply {
                        setToOrtho(true, width, height)
                    }
                )
            )
        }

        Gdx.app.logLevel = if (isDebug) Application.LOG_DEBUG else Application.LOG_ERROR
    }

    fun newGame(gameMode: Int) {
        setScreen(mainComponent.gameScreen.apply { newGame(gameMode) })
    }

    override fun newGameCreative() {
        newGame(1)
    }

    override fun newGameSurvival() {
        newGame(0)
    }

    override fun loadGame() {
        setScreen(mainComponent.gameScreen.apply { loadGame() })
    }

    override fun quitGame() {
        screen?.dispose()
        setScreen(mainComponent.menuScreen.apply { resetMenu() })
    }

    override fun exitGame() {
        Gdx.app.exit()
    }

    override fun create() {
        Gdx.files.absolute(gameDataDirectoryPath).mkdirs()
        initMainConfig()

        mainComponent.initializeAssetsUseCase()
        setScreen(mainComponent.menuScreen)
    }

    override fun dispose() {
        screen?.dispose()
        mainComponent.disposeAssetsUseCase()
    }


    companion object {
        private const val TAG = "CaveGame"
        private const val DEFAULT_VIEWPORT_WIDTH = 480f
    }

}